### ISSUE TITLE
In-batch-negatives and distributed traning

### ISSUE THREAD

**Issue by mcschmitz** [tensorflow/tensorflow#47344]
I'm training a model using triplet loss. This means that my loss function aims to minimize the distance to a positive observation (let's say a sentence with the same semantic meaning) and maximize the distance to the negative observation. 
Im using [TripletHardLoss](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/triplet.py#L348) from tensorflow-addons to do so. TripletHardLoss mines the batch to find the positive observation with the largest distance and the negative observation with the smallest distance in the batch and then applies the triplet loss.

Now when I train this model using `tf.distribute.MirroredStrategy` on a batch of 128 sentences and 8 GPUs the batches on my devices will only contain 14 (not counting the positive observation and the observation itself) potential negatives for each datapoint in the batch.

Is there any distribution strategy to concatenate all the 128 embeddings, run the tripling mining on this global batch, distribute the power observation losses back to the devices and calculate the gradients (or even calculate the gradients on the device that runs the triplet mining) to utilize the negatives of the global batch?

**Comment by mcschmitz** [comment]
So I was kind of able to overcome this problem locally by manipulating the `train_step` method of the Model

```python3
def train_step(self, data):
    """
    Performs a training step with the given batch.

    Args:
        data: batch data
    """
    data = data_adapter.expand_1d(data)
    x, y, sample_weight = data_adapter.unpack_x_y_sample_weight(data)
    x = x if isinstance(x, tuple) else [x]
    y = y if isinstance(y, tuple) else [y]
    sample_weight = sample_weight if isinstance(x, tuple) else [sample_weight]

    batch_size = x[0].shape[0]
    minibatch_size = batch_size // self.gradient_accumulation_steps
    # self.gradient_accumulation_steps has been assigned when initializing the model.

    train_vars = self.trainable_variables
    accum_gradient = [tf.zeros_like(v) for v in train_vars]
    y_pred = self(x, training=True)
    # Predict embeddings for the entire batch
    for step in range(self.gradient_accumulation_steps):
        start = step * minibatch_size
        end = start + minibatch_size
        with backprop.GradientTape() as tape:
            _y_pred_i = self([xi[start:end] for xi in x], training=True)
            # Predict embeddings for the minibatch to record the gradients
            loss = self.compiled_loss(
                y,
                y_pred_i_full,
                sample_weight,
                regularization_losses=self.losses,
            )[start:end]
            # Calculate loss for the minibatch
        gradients = tape.gradient(loss, train_vars)
        accum_gradient = [
            (a + g) if g is not None else a
            for a, g in zip(accum_gradient, gradients)
        ]
    accum_gradient = [
        grad / self.gradient_accumulation_steps for grad in accum_gradient
    ]
    self.optimizer.apply_gradients(zip(accum_gradient, train_vars))

    self.compiled_metrics.update_state(y, y_pred, sample_weight)
    return {m.name: m.result() for m in self.metrics}
```

It's a bit hacky though and while it allows me to do triplet mining with a large batch on my local machine I'm not sure whether it solves the distribution problem. So I'd still appreciate any recommendation or tip on how to solve this in a more efficient way.

**Comment by yuefengz** [comment]
If I understand it correctly, what you can do is before loss computation, use `tf.distribute.ReplicaContext.all_gather` to gather the logits from all the replicas and compute a loss based on the global logits.
